tests/test_read_lazy.py

```diff
@@ -0,0 +1,592 @@
+        #  TODO:xr.merge always upcasts to float due to NA and you can't downcast?
+        for col in remote_df_corrected.columns:
+            dtype = remote_df_corrected[col].dtype
+            if dtype in [np.float64, np.float32]:
+                var_df[col] = var_df[col].astype(dtype)
+        assert_equal(remote_df_corrected, var_df)
+        for key in var_keys_to_track:
+            stores[store_idx].assert_access_count(key, 1)
+            for store in stores:
+                if store != stores[store_idx]:
+                    store.assert_access_count(key, 0)
+        stores[store_idx].reset_key_trackers()
+
+
+@pytest.mark.parametrize(
+    "index",
+    [
+        pytest.param(
+            slice(500, 1500),
+            id="slice",
+        ),
+        pytest.param(
+            np.arange(950, 1050),
+            id="consecutive integer array",
+        ),
+        pytest.param(
+            np.random.choice(np.arange(800, 1100), 500),
+            id="random integer array",
+        ),
+        pytest.param(
+            np.random.choice([True, False], 2000),
+            id="boolean array",
+        ),
+        pytest.param(slice(None), id="full slice"),
+        pytest.param("a", id="categorical_subset"),
+        pytest.param(None, id="No index"),
+    ],
+)
+def test_concat_full_and_subsets(
+    adata_remote_orig: tuple[AnnData, AnnData],
+    join: Join_T,
+    index: slice | NDArray | Literal["a"] | None,
+    load_annotation_index: bool,
+):
+    from anndata.experimental.backed._compat import Dataset2D
+
+    remote, orig = adata_remote_orig
+
+    maybe_warning_context = (
+        pytest.warns(UserWarning, match=r"Concatenating with a pandas numeric")
+        if not load_annotation_index
+        else nullcontext()
+    )
+    with maybe_warning_context:
+        remote_concatenated = ad.concat([remote, remote], join=join)
+    if index is not None:
+        if np.isscalar(index) and index == "a":
+            index = remote_concatenated.obs["obs_cat"] == "a"
+        remote_concatenated = remote_concatenated[index]
+    orig_concatenated = ad.concat([orig, orig], join=join)
+    if index is not None:
+        orig_concatenated = orig_concatenated[index]
+    in_memory_remote_concatenated = remote_concatenated.to_memory()
+    corrected_remote_obs, corrected_memory_obs = correct_extension_dtype_differences(
+        in_memory_remote_concatenated.obs, orig_concatenated.obs
+    )
+    assert isinstance(remote_concatenated.obs, Dataset2D)
+    assert_equal(corrected_remote_obs, corrected_memory_obs)
+    assert_equal(in_memory_remote_concatenated.X, orig_concatenated.X)
+    assert (
+        in_memory_remote_concatenated.var_names.tolist()
+        == orig_concatenated.var_names.tolist()
+    )
+
+
+@pytest.mark.parametrize(
+    "elem_key",
+    map(
+        lambda x: pytest.param(x, id="-".join(map(str, x))),
+        [("obs", None), ("var", None), ("obsm", "df"), ("varm", "df")],
+    ),
```

Comment:
please use comprehensions instead of mapping lambdas.

also please unpack the tuple into two parameters here instead of doing `elem_key[0|1]` a bunch, something like: `("attr", "key"), [pytest.param(attr, key, id=...) for attr, key in [...]]`